tokio-macros 1.8.1 regression

NOTE(taiki-e): I yanked 1.8.1 for now.

---

I have a suspicion that this cause an interesting regression, but I'm not certain.

In our codebase, we have something along the lines of
```rust
macro_rules! custom_test {
    (async fn $name:ident() $b:block) => {
        #[::tokio::test]
        async fn $name() {
            dbg!("foo");
            $b
        }
    };
}

// usage normally with macro_rules_attribute, but for demonstration
custom_test!(
    async fn foo() {
        dbg!("bar");
    }
);
```

This compiles with tokio-macros 1.8.0 but does not with 1.8.1

```rust
error[E0425]: cannot find value `body` in this scope
  --> tests/main.rs:74:20
   |
74 |       async fn foo() {
   |  ____________________^
75 | |         dbg!("bar");
76 | |     }
   | |_____^ not found in this scope
```

_Originally posted by @conradludgate in https://github.com/tokio-rs/tokio/issues/5092#issuecomment-1331009348_

--- a/tokio-macros/src/entry.rs
+++ b/tokio-macros/src/entry.rs
@@ -383,6 +383,7 @@ fn parse_knobs(mut input: syn::ItemFn, is_test: bool, config: FinalConfig) -> To
 
     let body = &input.block;
     let brace_token = input.block.brace_token;
+    let body_ident = quote! { body };
     let block_expr = quote_spanned! {last_stmt_end_span=>
         #[allow(clippy::expect_used, clippy::diverging_sub_expression)]
         {
@@ -390,7 +391,7 @@ fn parse_knobs(mut input: syn::ItemFn, is_test: bool, config: FinalConfig) -> To
                 .enable_all()
                 .build()
                 .expect("Failed building the Runtime")
-                .block_on(body);
+                .block_on(#body_ident);
         }
     };
 
@@ -479,7 +480,7 @@ pub(crate) fn test(args: TokenStream, item: TokenStream, rt_multi_thread: bool) 
     };
     let config = if let Some(attr) = input.attrs.iter().find(|attr| attr.path.is_ident("test")) {
         let msg = "second test attribute is supplied";
-        Err(syn::Error::new_spanned(&attr, msg))
+        Err(syn::Error::new_spanned(attr, msg))
     } else {
         AttributeArgs::parse_terminated
             .parse(args)

--- a/tokio/tests/macros_test.rs
+++ b/tokio/tests/macros_test.rs
@@ -71,3 +71,18 @@ pub mod clippy_semicolon_if_nothing_returned {
         // To trigger clippy::semicolon_if_nothing_returned lint, the block needs to contain newline.
     }
 }
+
+// https://github.com/tokio-rs/tokio/issues/5243
+pub mod issue_5243 {
+    macro_rules! mac {
+        (async fn $name:ident() $b:block) => {
+            #[::tokio::test]
+            async fn $name() {
+                $b
+            }
+        };
+    }
+    mac!(
+        async fn foo() {}
+    );
+}

